src/core_plugins/vega_vis/public/vega_request_handler.js

```diff
@@ -0,0 +1,9 @@
+export function VegaRequestHandlerProvider() {
+  return {
+    name: 'vega',
+    handler: (/*vis, appState, uiState, queryFilter*/) => {
+      // return different data when handler is called so vega visualization can watch visData
```

Comment:
:) clever!

This is probably an artifact of bringin in `vega_view` and reusing that as-is. But this request-handler should actually run the actual query, basically what the `vega_loader` is doing right now.

I know this is going to somewhat mess up how you implemented the loader right now, it's basically something you inject into `vega.View`. So the `vega_loader` should just return whatever came out of the `request_handler`.

I'm not 100% familiar with Vega, so let me know how feasible this is. 

In any case, keeping the data-fetch OUTSIDE of the request_handler is going to hurt is in the long run. The `request_handler` is our funnel for visualizations to request data from the server (implied of course that 99% of requests end up hitting an Elasticsearch cluster, this is obviously not the case for mashing up data from 3rd party sources). It's important that ES-searchs are handled by requestHandlers because it is consistent with all our other visualizations.

Reply:
Hmm... this might be tricky, but TBH i'm not very happy with the current implementation either.

The `data` section is a list of data sources, and each could be an ES request or a simple URL. Vega calls the loader once for each data source, and expects a promise of the data.  I would like to use [Vega observable pattern](https://github.com/vega/vega/issues/893), possibly implementing it myself - which would remove the need to regenerate the graph on auto refresh.

One option might be to dynamically change the Vega spec with a pre-parser - inspect data section, and replace "url" with the hardcoded values right before passing it to Vega parser. We can even use `"query": {...}` instead of `"url": {...}`.  On auto-refresh, the data could be auto-re-injected into the Vega view.  Not sure how stable this approach would be...

Reply:
Not familiar enough with Vega to speak to the pre-parser.

Could we inject a callback in the vega_loader that returns the result from the requesthandler (or even the actual data-object)? We know where the ES-search query is on the Vega-object. The data-fetch that is now in vega_loader could be transfered to requestHandler. The instantiation of the vega.View happens after the requestHandler-call is made I believe, so the vega_loader can use that data from the requestHandler as-is.